src/man/pkglint.1
.\" Copyright (c) 2007, 2019, Oracle and/or its affiliates. All rights reserved.
.\" Copyright 2022 OmniOS Community Edition (OmniOSce) Association.
.Dd February 17, 2022
.Dt PKGLINT 1
.Os
.Sh NAME
.Nm pkglint
.Nd Image Packaging System package lint
.Sh SYNOPSIS
.Nm
.Op Fl c Ar cache_dir
.Oo Fl r Ar repo_uri Oc Ns \&...
.Op Fl p Ar regexp
.Op Fl e Ar extension_path
.Op Fl f Ar config_file
.Op Fl b Ar build_no
.Op Fl v
.Oo Fl l Ar lint_uri Oc Ns \&...
.Ar manifest Ns \&...
.Pp
.Nm
.Fl L
.Op Fl v
.Sh DESCRIPTION
.Nm
runs a series of checks on one or more package manifests, optionally
referencing another repository.
.Pp
.Nm
should be used during the package authoring process, prior to package
publication.
.Nm
performs exhaustive testing on the manifests that might be too expensive to
perform during normal operation of
.Xr pkgsend 1
or
.Xr pkg.depotd 8 .
.Nm
checks include tests for duplicate actions, missing attributes, and unusual
file permissions.
.Pp
Manifests for linting can be passed as a space-separated list of local files on
the command line, or manifests can be retrieved from a repository.
.Pp
When retrieving manifests from repositories, on first run
.Nm
creates and populates
.Xr pkg 7
user images in the specified cache directory.
If the
.Fl r
option is supplied, a user image named
.Ar cache_dir Ns Sy /ref_image
is created for the reference repository.
If the
.Fl l
option is supplied, a user image named
.Ar cache_dir Ns Sy /lint_image
is created for the lint repository.
No content is installed in these images.
These images are only used by
.Nm
to retrieve manifests from the repositories.
.Pp
Subsequent invocations of
.Nm
can reuse the cache directory and can omit any
.Fl r
or
.Fl l
arguments.
.Pp
.Nm
provides limited support for configuring publishers in the cache directory.
Use
.Xr pkg 1
to perform more complex publisher configuration on these images.
.Pp
.Nm
allows package authors to bypass checks for a given manifest or action.
A manifest or action that contains the attribute
.Sy pkg.linted
set to
.Sy True
does not produce any lint output for that manifest or action.
.Pp
More granular
.Sy pkg.linted
settings can be made using substrings of
.Nm
check names.
For example,
.Sy pkg.linted\&. Ns Em check Ns \&. Ns Em id
set to
.Sy True
bypasses all checks with the name
.Ar check Ns \&. Ns id
for the given manifest or action.
.Pp
The behaviour of
.Nm
can be configured by specifying a
.Sy pkglintrc
file.
By default,
.Nm
searches in
.Pa /usr/share/lib/pkg/pkglintrc
and
.Pa $HOME/.pkglintrc
for configuration options.
Use the
.Fl f
option to specify a different configuration file.
.Pp
During the lint run, any errors or warnings encountered are printed to
.Dv stderr .
.Sh OPTIONS
The following options are supported:
.Bl -tag -width Ds
.It Fl h , \&-help
Display a usage message.
.It Fl b Ar build_no
Specify a build number used to narrow the list of packages used during linting
from lint and reference repositories.
If no
.Fl b
option is specified, the latest versions of packages are used.
See also the
.Sy version.pattern
configuration property.
.It Fl c Ar cache_dir
Specify a local directory used for caching package metadata from the lint and
reference repositories.
.It Fl l Ar lint_uri
Specify a URI representing the location of the lint repository.
Both HTTP and file system based publication are supported.
If you specify
.Fl l ,
then you must also specify
.Fl c .
The
.Fl l
option can be specified multiple times.
.It Fl L
List the known and excluded lint checks and then exit.
Display the short name and description of each check.
When combined with the
.Fl v
flag, display the method that implements the check instead of the description.
.It Fl e Ar extension_path
Directory to add to the Python search path when loading
.Nm
extensions.
.It Fl f Ar config_file
Configure the
.Nm
session using the
.Ar config_file
configuration file.
.It Fl p Ar regexp
Specify a regular expression used to narrow the list of packages to be checked
from the lint repository.
All manifests from the reference repository are loaded
.Po
assuming they match the value for
.Fl b ,
if supplied
.Pc ,
ignoring this pattern.
.It Fl r Ar repo_uri
Specify a URI representing the location of the reference repository.
If you specify
.Fl r ,
then you must also specify
.Fl c .
The
.Fl r
option can be specified multiple times.
.It Fl v
Run
.Nm
in a verbose mode, overriding any
.Sy log_level
settings in the configuration file.
.El
.Sh FILES
The
.Sy pkglintrc
configuration file takes the following key/value arguments:
.Bl -tag -width Ds
.It Sy log_level
The minimum level at which to emit lint messages.
Lint messages lower than this level are discarded.
The default value is
.Sy INFO .
.Pp
Log levels in order of least to most severe are
.Sy DEBUG ,
.Sy INFO ,
.Sy WARNING ,
.Sy ERROR ,
and
.Sy CRITICAL .
.It Sy do_pub_checks
If
.Sy True ,
perform checks that might only make sense for published packages.
The default value is
.Sy True .
.It Sy pkglint.ext.*
The plugin mechanism of
.Nm
allows for additional lint modules to be added at runtime.
Any key that starts with
.Sy pkglint.ext\&.
takes a value that must be a fully-specified Python module.
See the
.Sx Developers
section for more information.
The
.Ar extension_path
option specifies additional directory locations to search when loading
extensions.
.It Sy extension_path
A list of directories, separated by
.Sq \&: ,
to search for extension modules.
If the
.Fl e
flag is specified on the command line, it is added higher in the search order
than this configuration file option.
.It Sy pkglint.exclude
A space-separated list of fully specified Python modules, classes, or function
names to omit from the set of checks performed.
.It Sy use_progress_tracker
If
.Sy True ,
use a progress tracker when iterating over manifests during lint runs.
The default value is
.Sy True .
.It Sy version.pattern
A version pattern, used when specifying a build number to lint against
.Pq Fl b .
If not specified in the configuration file, the
.Fl b
option uses the pattern
.Sy *,5.11-0\&. ,
matching all components of the 5.11 build, with a branch prefix of 0.
.El
.Sh DEVELOPERS
To extend the set of checks performed by
.Nm ,
subclass
.Sy pkg.lint.base.Checker
and its subclasses,
.Sy ManifestChecker
and
.Sy ActionChecker .
Add the Python module name that contains those classes to a new
.Sy pkglint.ext\&.
key in the configuration file.
.Pp
Instances of those new subclasses are created by
.Nm
on startup.
Methods inside each subclass with the special keyword argument
.Sy pkglint_id
are invoked during the course of the lint session.
Those methods should have the same signature as the corresponding
.Sy check()
method in the super class.
Methods should also be assigned a
.Sy pkglint_desc
attribute, which is used as the description printed by
.Sy pkglint -L .
.Pp
Parameters are available to
.Sy Checker
subclasses, allowing them to tune their behaviour.
The recommended parameter naming convention is
.Em pkglint_id Ns \&. Ar name .
Parameter values can be stored in the configuration file, or accessed in
manifests or actions retrieved using the
.Sy LintEngine.get_param()
method.
When accessing parameters from the manifest, the prefix
.Sy pkg.lint
is prepended to the key name to ensure that
.Nm
parameters do not overlap with any existing action or manifest values.
.Sh EXIT STATUS
The following exit values are returned:
.Bl -tag -width Ds
.It 0
Command succeeded.
.It 1
One or more package manifests contain lint errors.
.It 2
An error occurred that is not a lint error in a manifest.
For example, an invalid command line option might have been specified.
.It 99
An unanticipated exception occurred.
.El
.Sh EXAMPLES
.Sy Example 1 No First Run on a Particular Repository
.Pp
Running a
.Nm
session for the first time on a given repository.
.Bd -literal -offset 4n
pkglint -c /space/cache -r http://localhost:10000 mymanifest.mf
.Ed
.Pp
.Sy Example 2 No Subsequent Run on the Same Repository
.Pp
A subsequent run against the same repository used in Example 1.
.Bd -literal -offset 4n
pkglint -c /space/cache mymanifest-fixed.mf
.Ed
.Pp
.Sy Example 3 No Using a Lint Repository With a Narrowed Manifest Set
.Pp
Running a
.Nm
session with a lint repository and specifying a subset of manifests to check.
.Bd -literal -offset 4n
pkglint -c /space/othercache -l http://localhost:10000 \e
    -p '.*firefox.*'
.Ed
.Pp
.Sy Example 4 No Specifying a Build
.Pp
Running a
.Nm
session against a given build in verbose mode.
.Bd -literal -offset 4n
pkglint -c /space/cache -r http://localhost:10000 \e
    -l http://localhost:12000 -b 147 -v
.Ed
.Pp
.Sy Example 5 No Modifying a Configuration File
.Pp
A configuration file with a new lint module, excluding some checks.
.Bd -literal -offset 4n
cat ~/.pkglintrc
  [pkglint]
  log_level = DEBUG
  # log_level = INFO
  pkglint.ext.mycheck = org.timf.mychecks
  pkglint.ext.opensolaris = pkg.lint.opensolaris
  pkglint.exclude: pkg.lint.opensolaris.OpenSolarisActionChecker
  pkg.lint.pkglint.PkgActionChecker.unusual_perms
  pkg.lint.pkglint.PkgManifestChecker
  pkg.lint.opensolaris.OpenSolarisManifestChecker
.Ed
.Sh INTERFACE STABILITY
The command line interface of
.Nm
is
.Sy Uncommitted .
The output of
.Nm
is
.Sy Not-An-Interface
and may change at any time.
.Sh SEE ALSO
.Xr pkg 1 ,
.Xr pkgsend 1 ,
.Xr pkg 7 ,
.Xr pkg.depotd 8


doc/history.txt

pkg
HISTORY

1. Summary

The intent of history is to enable pkg clients to maintain a record of all
image-modifying operations that they perform.

2. Discussion

2.1 Operations

    The following operations are currently recorded by the clients that are
    part of pkg(7):

        add-publisher
        image-create
        image-set-attributes
        image-update
        install
        purge-history
        rebuild-index
        refresh-publisher
        remove-publisher
        set-publisher
        set-preferred-publisher
        uninstall
        update-publisher

    These operations were chosen because they have the potential to alter the
    behavior of pkg(7) operations or because they modify an image.

3. History Entries

3.1 Information Recorded

    Each history entry contains a record of basic information about the client
    performing the operation, along with a mix of required and optional
    information about the operation performed.

    The following information about the client is always recorded:

        client_name     - The name of the client performing the operation
                          (e.g. pkg, packagemanager)

        client_version  - The value of pkg.VERSION at the time the client
                          performed the operation (e.g. 2e5300c4f0a4+)

        client_args     - The command-line arguments used when executing the
                          client (e.g. /usr/bin/pkg install foo)

    The following information about the operation performed is always recorded:

        operation_name  - The name of the operation performed (refer to 2.1)

        start_time      - When the operation started (e.g. 20080916T214726Z)

        end_time        - When the operation ended (e.g. 20080916T214800Z)

        userid          - The id of the user that performed the operation
                          (e.g. 0)

        username        - The username of the user that performed the operation
                          (e.g. root)

        result          - The outcome of the operation and the reason for it
                          (e.g. Failed, Transport) This maps to the pkg history
                          output columns of 'outcome' and 'reason'.

    The following information about the operation will be recorded if provided
    by the client or api:

        start_state     - Information about the operation requested before any
                          evaluation of that request is performed (e.g. an
                          image plan before evaluation)

        end_state       - Information about the operation requested after
                          evaluation of that request has been performed (e.g.
                          image plan after evaluation)

        errors          - Any errors that were encountered while performing the
                          operation (i.e. tracebacks, exceptions, etc.)

        be              - The name of the boot environment on which the
                          operation was performed

        be_uuid         - The uuid corresponding to the boot environment

        new_be          - The name of any new boot environment that was created
                          while performing the operation.

        new_be_uuid     - The uuid corresponding to the new boot environment

        snapshot        - The name of the snapshot that was taken as a result of
                          this operation.  If the operation completed
                          successfully and the snapshot was destroyed, this
                          this information is not stored.

3.2 Storage

    Each history entry is recorded in a XML file located in the metadata
    directory of an image (e.g. /var/pkg) in a directory named "history".
    Each XML file is named after the pattern %Y%m%dT%H%M%SZ-sequence.xml.
    Where sequence is a numeric value appended so that when multiple
    operations are performed on an image within the same second (e.g. -02,
    -03, etc.) entries are still written correctly.

3.3 File Format

    It should be noted that this format description is that of a private
    interface and is subject to change.

    History entries are recorded in a XML file with a fairly simplistic
    structure.  The format is as follows:

        The first line of each file is a standard xml header along with the
        encoding used to save the information.  This allows the pkg history
        command to correctly display this information if the client changes
        locales later.

        <?xml version="1.0" encoding="ascii"?>

        The second element of every history XML file is a root element used to
        contain the client and operation information.  Only one per file ever
        occurs.

        <history>

            The client element has a name and version attribute used to record
            client_name and client_version.  Only one per file ever occurs.

            <client name="pkg" version="2e5300c4f0a4+">

                The args element contains one or more "arg" (argument) elements
                that match each element of the system argument list used to
                execute the client.  It has no attributes.  Only one per file
                ever occurs.

                <args>

                    Each arg element contains a CDATA element to encapsulate
                    the raw information for the argument.  It has no attributes.

                    <arg>
                        <![CDATA[/usr/bin/pkg]]>
                    </arg>
                    <arg>
                        <![CDATA[-R]]>
                    </arg>
                    <arg>
                        <![CDATA[/tmp/test-image]]>
                    </arg>
                    <arg>
                        <![CDATA[install]]>
                    </arg>
                    <arg>
                        <![CDATA[SUNWvim]]>
                    </arg>
                </args>
            </client>


src/util/publish/pkglint.py
#!/usr/bin/python3 -Es
#
# CDDL HEADER START
#
# The contents of this file are subject to the terms of the
# Common Development and Distribution License (the "License").
# You may not use this file except in compliance with the License.
#
# You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
# or http://www.opensolaris.org/os/licensing.
# See the License for the specific language governing permissions
# and limitations under the License.
#
# When distributing Covered Code, include this CDDL HEADER in each
# file and include the License file at usr/src/OPENSOLARIS.LICENSE.
# If applicable, add the following below this CDDL HEADER, with the
# fields enclosed by brackets "[]" replaced with your own identifying
# information: Portions Copyright [yyyy] [name of copyright owner]
#
# CDDL HEADER END
#

#
# Copyright (c) 2010, 2019, Oracle and/or its affiliates. All rights reserved.
#

import pkg.site_paths; pkg.site_paths.init()
import codecs
import logging
import six
import sys
import os
import gettext
import locale
import traceback
import warnings
from optparse import OptionParser

from pkg.client.api_errors import InvalidPackageErrors
from pkg import VERSION
from pkg.misc import PipeError
from pkg.client.pkgdefs import EXIT_OK, EXIT_OOPS, EXIT_BADOPT

import pkg.lint.engine as engine
import pkg.lint.log as log
import pkg.fmri as fmri
import pkg.manifest
import pkg.misc as misc
import pkg.client.api_errors as apx
import pkg.client.transport.exception as tx

logger = None

def error(message=""):
        """Emit an error message prefixed by the command name. """
        misc.emsg("pkglint: {0}".format(message))

        if logger is not None:
                logger.error(_("Error: {0}").format(message))

def msg(message):
        logger.info(message)

def debug(message):
        logger.debug(message)

def main_func():
        """Start pkglint."""

        global logger

        usage = \
            _("\n"
            "        %prog [-b build_no] [-c cache_dir] [-f file]\n"
            "            [-l uri ...] [-p regexp] [-r uri ...] [-v]\n"
            "            [-e extension_path ...]\n"
            "            manifest ...\n"
            "        %prog -L")
        parser = OptionParser(usage=usage)

        parser.add_option("-b", dest="release", metavar="build_no",
            help=_("build to use from lint and reference repositories"))
        parser.add_option("-c", dest="cache", metavar="dir",
            help=_("directory to use as a repository cache"))
        parser.add_option("-f", dest="config", metavar="file",
            help=_("specify an alternative pkglintrc file"))
        parser.add_option("-l", dest="lint_uris", metavar="uri",
            action="append", help=_("lint repository URI"))
        parser.add_option("-L", dest="list_checks",
            action="store_true",
            help=_("list checks configured for this session and exit"))
        parser.add_option("-p", dest="pattern", metavar="regexp",
            help=_("pattern to match FMRIs in lint URI"))
        parser.add_option("-r", dest="ref_uris", metavar="uri",
            action="append", help=_("reference repository URI"))
        parser.add_option("-e", dest="extension_path", metavar="dir",
            action="append", help=_("extension_path"))
        parser.add_option("-v", dest="verbose", action="store_true",
            help=_("produce verbose output, overriding settings in pkglintrc"))

        opts, args = parser.parse_args(sys.argv[1:])

        # without a cache option, we can't access repositories, so expect
        # local manifests.
        if not (opts.cache or opts.list_checks) and not args:
                parser.error(
                    _("Required -c option missing, no local manifests provided."
                    ))

        pattern = opts.pattern
        opts.ref_uris = _make_list(opts.ref_uris)
        opts.lint_uris = _make_list(opts.lint_uris)

        logger = logging.getLogger("pkglint")
        ch = logging.StreamHandler(sys.stdout)

        if opts.verbose:
                logger.setLevel(logging.DEBUG)
                ch.setLevel(logging.DEBUG)

        else:
                logger.setLevel(logging.INFO)
                ch.setLevel(logging.INFO)

        logger.addHandler(ch)

        lint_logger = log.PlainLogFormatter()
        try:
                if not opts.list_checks:
                        msg(_("Lint engine setup..."))
                lint_engine = engine.LintEngine(lint_logger,
                    config_file=opts.config, verbose=opts.verbose,
                    extension_path=opts.extension_path)

                if opts.list_checks:
                        list_checks(lint_engine.checkers,
                            lint_engine.excluded_checkers, opts.verbose)
                        return EXIT_OK

                if (opts.lint_uris or opts.ref_uris) and not opts.cache:
                        parser.error(
                            _("Required -c option missing when using "
                            "repositories."))

                manifests = []
                if len(args) >= 1:
                        manifests = read_manifests(args, lint_logger)
                        if None in manifests or \
                            lint_logger.produced_lint_msgs():
                                error(_("Fatal error in manifest - exiting."))
                                return EXIT_OOPS
                lint_engine.setup(ref_uris=opts.ref_uris,
                    lint_uris=opts.lint_uris,
                    lint_manifests=manifests,
                    cache=opts.cache,
                    pattern=pattern,
                    release=opts.release)

                msg(_("Starting lint run..."))

                lint_engine.execute()
                lint_engine.teardown()
                lint_logger.close()

        except engine.LintEngineSetupException as err:
                # errors during setup are likely to be caused by bad
                # input or configuration, not lint errors in manifests.
                error(err)
                return EXIT_BADOPT

        except engine.LintEngineException as err:
                error(err)
                return EXIT_OOPS

        if lint_logger.produced_lint_msgs():
                return EXIT_OOPS
        else:
                return EXIT_OK

def list_checks(checkers, exclude, verbose=False):
        """Prints a human-readable version of configured checks."""

        # used for justifying output
        width = 28

        def get_method_desc(method, verbose):
                if "pkglint_desc" in method.__dict__ and not verbose:
                        return method.pkglint_desc
                else:
                        return "{0}.{1}.{2}".format(method.__self__.__class__.__module__,
                            method.__self__.__class__.__name__,
                            method.__func__.__name__)

        def emit(name, value):
                msg("{0} {1}".format(name.ljust(width), value))

        def print_list(items):
                k = list(items.keys())
                k.sort()
                for lint_id in k:
                        emit(lint_id, items[lint_id])

        include_items = {}
        exclude_items = {}

        for checker in checkers:
                for m, lint_id in checker.included_checks:
                        include_items[lint_id] = get_method_desc(m, verbose)

        for checker in exclude:
                for m, lint_id in checker.excluded_checks:
                        exclude_items[lint_id] = get_method_desc(m, verbose)
                for m, lint_id in checker.included_checks:
                        exclude_items[lint_id] = get_method_desc(m, verbose)

        for checker in checkers:
                for m, lint_id in checker.excluded_checks:
                        exclude_items[lint_id] = get_method_desc(m, verbose)

        if include_items or exclude_items:
                if verbose:
                        emit(_("NAME"), _("METHOD"))
                else:
                        emit(_("NAME"), _("DESCRIPTION"))
                print_list(include_items)

                if exclude_items:
                        msg(_("\nExcluded checks:"))
                        print_list(exclude_items)

def read_manifests(names, lint_logger):
        """Read a list of filenames, return a list of Manifest objects."""

        manifests = []
        for filename in names:
                data = None
                # borrowed code from publish.py
                lines = []      # giant string of all input lines
                linecnts = []   # tuples of starting line no., ending line no
                linecounter = 0 # running total
                try:
                        f = codecs.open(filename, "rb", "utf-8")
                        data = f.read()
                except UnicodeDecodeError as e:
                        lint_logger.critical(_("Invalid file {file}: "
                            "manifest not encoded in UTF-8: {err}").format(
                            file=filename, err=e),
                            msgid="lint.manifest002")
                        continue
                except IOError as e:
                        lint_logger.critical(_("Unable to read manifest file "
                            "{file}: {err}").format(file=filename, err=e),
                            msgid="lint.manifest001")
                        continue
                lines.append(data)
                linecnt = len(data.splitlines())
                linecnts.append((linecounter, linecounter + linecnt))
                linecounter += linecnt

                manifest = pkg.manifest.Manifest()
                try:
                        manifest.set_content(content="\n".join(lines))
                except pkg.actions.ActionError as e:
                        lineno = e.lineno
                        for i, tup in enumerate(linecnts):
                                if lineno > tup[0] and lineno <= tup[1]:
                                        lineno -= tup[0]
                                        break
                        else:
                                lineno = "???"

                        lint_logger.critical(
                            _("Error in {file} line: {ln}: {err} ").format(
                            file=filename,
                            ln=lineno,
                            err=str(e)), "lint.manifest002")
                        manifest = None
                except InvalidPackageErrors as e:
                        lint_logger.critical(
                            _("Error in file {file}: {err}").format(
                            file=filename,
                            err=str(e)), "lint.manifest002")
                        manifest = None

                if manifest and "pkg.fmri" in manifest:
                        try:
                                manifest.fmri = \
                                    pkg.fmri.PkgFmri(manifest["pkg.fmri"])
                        except fmri.IllegalFmri as e:
                                lint_logger.critical(
                                    _("Error in file {file}: "
                                    "{err}").format(
                                    file=filename, err=e),
                                    "lint.manifest002")
                        if manifest.fmri:
                                if not manifest.fmri.version:
                                        lint_logger.critical(
                                            _("Error in file {0}: "
                                            "pkg.fmri does not include a "
                                            "version string").format(filename),
                                            "lint.manifest003")
                                else:
                                        manifests.append(manifest)

                elif manifest:
                        lint_logger.critical(
                            _("Manifest {0} does not declare fmri.").format(filename),
                            "lint.manifest003")
                else:
                        manifests.append(None)
        return manifests

def _make_list(opt):
        """Makes a list out of opt, and returns it."""

        if isinstance(opt, list):
                return opt
        elif opt is None:
                return []
        else:
                return [opt]


if __name__ == "__main__":
        misc.setlocale(locale.LC_ALL, "", error)
        gettext.install("pkg", "/usr/share/locale",
            codeset=locale.getpreferredencoding())
        misc.set_fd_limits(printer=error)

        if six.PY3:
                # disable ResourceWarning: unclosed file
                warnings.filterwarnings("ignore", category=ResourceWarning)
        try:
                __ret = main_func()
        except (PipeError, KeyboardInterrupt):
                # We don't want to display any messages here to prevent
                # possible further broken pipe (EPIPE) errors.
                __ret = EXIT_BADOPT
        except SystemExit as __e:
                __ret = __e.code
        except (apx.InvalidDepotResponseException, tx.TransportFailures) as __e:
                error(__e)
                __ret = EXIT_BADOPT
        except:
                traceback.print_exc()
                error(misc.get_traceback_message())
                __ret = 99

        sys.exit(__ret)

# Vim hints
# vim:ts=8:sw=8:et:fdm=marker
